In [7]:
import pandas as pd
import numpy as np
from statsmodels.stats.outliers_influence import variance_inflation_factor


In [31]:
df = pd.read_csv('https://raw.githubusercontent.com/sanjeev49/Older-People-activity-prediction/main/raw_dio.csv')

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75128 entries, 0 to 75127
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Time            75128 non-null  float64
 1   Acc. Front      75128 non-null  float64
 2   Acc. vert       75128 non-null  float64
 3   Acc. Lat        75128 non-null  float64
 4   id              75128 non-null  int64  
 5   RSSI            75128 non-null  float64
 6   Phase           75128 non-null  float64
 7   Freq            75128 non-null  float64
 8   Activity Label  75128 non-null  int64  
dtypes: float64(7), int64(2)
memory usage: 5.2 MB


In [5]:
x= df.iloc[:, :-1]
y = df.iloc[:, -1]


In [8]:
arr = np.array(x)


In [14]:
arr[0]

array([ 0.0000e+00,  2.7203e-01,  1.0082e+00, -8.2102e-02,  1.0000e+00,
       -6.3500e+01,  2.4252e+00,  9.2425e+02])

In [19]:
vif2 = pd.DataFrame()
vif2['vif'] = [variance_inflation_factor(arr, i) for i in range(arr.shape[1])]

In [22]:
vif2['columns'] = x.columns
vif2

vif     columns
0    2.909534        Time
1   18.055771  Acc. Front
2    6.799554   Acc. vert
3    2.631783    Acc. Lat
4    6.570744          id
5  139.008119        RSSI
6    3.111670       Phase
7  165.823909        Freq

In [23]:
# Here are lot of outliers like in 'Acc. Front', 'RSSI', 'Freq'

In [24]:
# We treat  the outliers by IQR

In [25]:
def remove_outliers(df):
    Q1 = np.percentile(df, 25, interpolation='midpoint')
    Q3 = np.percentile(df,75, interpolation='midpoint')
    IQR = Q3-Q1
    upper=  np.where(df>=(Q3+1.5*IQR))
    lower = np.where(df<=(Q3-1.5*IQR))
    df.drop(upper[0],inplace=True,axis=0)
    df.drop(lower[0],inplace=True,axis=0)
    return df
Time = remove_outliers(df.Time)
front = remove_outliers(df['Acc. Front'])
vert = remove_outliers(df['Acc. vert'])
lat = remove_outliers(df['Acc. Lat'])
Rssi = remove_outliers(df.RSSI)
phase= remove_outliers(df.Phase)
Freq = remove_outliers(df.Freq)
    
        

In [26]:
df_2 = pd.DataFrame()
time = Time.loc[:52000]
front = front.iloc[:52000]
vert = vert.iloc[:52000]
lat = lat.iloc[:52000]
Rssi = Rssi.iloc[:52000]
phase = phase.iloc[:52000]
freq = Freq.iloc[:52000]

In [27]:
df_2['Time'] = time
df_2['Acc. Front'] = front
df_2['Acc. vert'] = vert
df_2['Acc. Lat'] = lat
df_2['id'] = df.id
df_2['RSSI']  = Rssi
df_2['Phase'] = phase
df_2['Freq']  = Freq

In [28]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52001 entries, 0 to 52000
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Time        52001 non-null  float64
 1   Acc. Front  51271 non-null  float64
 2   Acc. vert   51951 non-null  float64
 3   Acc. Lat    49855 non-null  float64
 4   id          52001 non-null  int64  
 5   RSSI        45988 non-null  float64
 6   Phase       52000 non-null  float64
 7   Freq        52001 non-null  float64
dtypes: float64(7), int64(1)
memory usage: 3.6 MB


In [29]:
# The Problem is after treating outliers the row counts of each column has been different. Prviously it was

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75128 entries, 0 to 75127
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Time            75128 non-null  float64
 1   Acc. Front      75128 non-null  float64
 2   Acc. vert       75128 non-null  float64
 3   Acc. Lat        75128 non-null  float64
 4   id              75128 non-null  int64  
 5   RSSI            75128 non-null  float64
 6   Phase           75128 non-null  float64
 7   Freq            75128 non-null  float64
 8   Activity Label  75128 non-null  int64  
dtypes: float64(7), int64(2)
memory usage: 5.2 MB


# As  you can see previously there are 75128 rows in each column . after removing outliers it is changed .
# How can i resolve this issue?